mypyc/freesymbols.py

```diff
@@ -0,0 +1,49 @@
+from typing import Dict, List, Set
+
+from mypy.nodes import FuncDef, NameExpr, SymbolNode, Var
+from mypy.traverser import TraverserVisitor
+
+
+class FreeSymbolsVisitor(TraverserVisitor):
+    """Class used to visit nested functions and determine free symbols."""
+    def __init__(self) -> None:
+        super().__init__()
+        self.free_symbols = {}  # type: Dict[FuncDef, Set[SymbolNode]]
+        self.symbols_to_fdefs = {}  # type: Dict[SymbolNode, FuncDef]
+        self.fdefs = []  # type: List[FuncDef]
```

Comment:
Document the meanings of the data structures. Notably, `free_symbols[f]` doesn't contain the variables that are free in `f` -- it contains the variables that are bound in `f` and occur free in nested functions inside `f`.